boards: esp32: Add support for M5Stack AtomS3

kartben opened the pull request:
This introduces support for the ESP32S3 based M5Stack AtomS3.

Tested all of the following:

- [x] LCD display and accelerometer tested with `samples/modules/lvgl/accelerometer_chart`, which works out-of-the-box (see below!)
- [x] Button (`samples/basic/button` and `samples/subsys/display/lvgl` sample)
- [x] Bluetooth (`samples/bluetooth/peripheral_hr`)
- [x] Wi-Fi (`samples/net/wifi`, wifi scan works)
- [x] Grove connector, tested with `samples/sensor/sensor_shell` and an I2C device attached

Doc page: https://builds.zephyrproject.io/zephyr/pr/64434/docs/boards/xtensa/m5stack_atoms3/doc/index.html

[Pxl 20231026 101330936.webm](https://github.com/zephyrproject-rtos/zephyr/assets/128251/d233b3cb-3746-4376-accc-271346083388)

kartben committed:
boards: esp32: Add support for M5Stack AtomS3

This introduces support for the ESP32S3 based M5Stack AtomS3.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben commented:
The first two commits are part of PR #64451 so you may ignore them.

sylvioalves commented:
@kartben awesome!

kartben commented:
Rebased as the 2 commits mentioned earlier are not needed anymore now that 1fac5ed2a60b5365953e74c6dbc1ba3640b728ad fixed the issue with Xtensa CI.

kartben commented:
@sylvioalves should I exclude this board from tests/lib/heap testcase, like other (but not all??) esp32-s2 and s3? I don't think the test failed in earlier runs though, but not 100% sure.

kartben commented:
@sylvioalves @marekmatej what should I be doing with the failing heap test? should I just add this board as a `platform_exclude` in the testcase? There are other ESP32S3 that are working just fine (but also some that are excluded).

```
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /__w/zephyr/zephyr/twister-out/m5stack_atoms3/tests/lib/heap/libraries.heap/zephyr/zephyr_pre0.map 
: && ccache /opt/toolchains/zephyr-sdk-0.16.3/xtensa-espressif_esp32s3_zephyr-elf/bin/xtensa-espressif_esp32s3_zephyr-elf-gcc  -gdwarf-4 zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/xtensa/core/offsets/offsets.c.obj  -fuse-ld=bfd  -T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/__w/zephyr/zephyr/twister-out/m5stack_atoms3/tests/lib/heap/libraries.heap/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/xtensa/core/libarch__xtensa__core.a  zephyr/lib/libc/picolibc/liblib__libc__picolibc.a  zephyr/lib/libc/common/liblib__libc__common.a  zephyr/lib/posix/liblib__posix.a  zephyr/subsys/random/libsubsys__random.a  zephyr/subsys/testsuite/ztest/libsubsys__testsuite__ztest.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/entropy/libdrivers__entropy.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  zephyr/drivers/regulator/libdrivers__regulator.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/opt/toolchains/zephyr-sdk-0.16.3/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0"  -L/__w/zephyr/zephyr/twister-out/m5stack_atoms3/tests/lib/heap/libraries.heap/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -lgcc  -Wl,--fatal-warnings  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -Wl,-no-pie  -DPICOLIBC_INTEGER_PRINTF_SCANF  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/src/linker/esp32s3.rom.alias.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.api.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.libgcc.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.newlib.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.version.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/soc/esp32s3/ld/esp32s3.peripherals.ld  -Wl,--wrap=k_calloc  -Wl,--wrap=k_malloc  --specs=picolibc.specs  -lc  -lgcc && cd /__w/zephyr/zephyr/twister-out/m5stack_atoms3/tests/lib/heap/libraries.heap/zephyr && /usr/bin/cmake -E true
/opt/toolchains/zephyr-sdk-0.16.3/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/zephyr_pre0.elf section `.dram0.data' will not fit in region `dram0_0_seg'
/opt/toolchains/zephyr-sdk-0.16.3/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: region `dram0_0_seg' overflowed by 1536 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

sylvioalves commented on boards/xtensa/m5stack_atoms3/m5stack_atoms3_defconfig:
```diff
@@ -0,0 +1,18 @@
+CONFIG_UART_CONSOLE=y
+
+# for debugging
+CONFIG_SHELL=y
```
I suggest removing this as default config. This will allow `test/lib/heap` to pass. User can manually enable this if necessary. Sounds good?

kartben replied:
Sure thing! Definitely the right thing to do. Pretty sure I copy pasted from another esp32 board fwiw, including the "for debugging" comment.

sylvioalves commented on boards/xtensa/m5stack_atoms3/m5stack_atoms3_defconfig:
```diff
@@ -0,0 +1,18 @@
+
+CONFIG_MAIN_STACK_SIZE=2048
+
+CONFIG_ESP_HEAP_MEM_POOL_REGION_1_SIZE=0
```
This is not needed nor available for ESP32S3. Only for ESP32 that has dual ram bank. So you can remove it from here.

sylvioalves commented:
> @sylvioalves should I exclude this board from tests/lib/heap testcase, like other (but not all??) esp32-s2 and s3? I don't think the test failed in earlier runs though, but not 100% sure.

If you remove the `CONGI_SHELL=y`, test will pass.

kartben commented:
> > @sylvioalves should I exclude this board from tests/lib/heap testcase, like other (but not all??) esp32-s2 and s3? I don't think the test failed in earlier runs though, but not 100% sure.
> 
> If you remove the `CONFIG_SHELL=y`, test will pass.

Thanks for your help! All green now!
Let me know who else you think could help reviewing this PR btw, as it hasn't had a lot of traction in the past two weeks. Thanks!

sylvioalves commented:
@marekmatej @LucasTambor PTAL

LucasTambor commented on boards/xtensa/m5stack_atoms3/support/openocd.cfg:
Is this file necessary? In the index.rst says:

```
M5Stack AtomS3 debugging is not supported due to pinout limitations.
```

kartben replied:
Ha! Good catch. Will remove.

kartben replied:
done, thx!

sylvioalves commented on boards/xtensa/m5stack_atoms3/m5stack_atoms3.yaml:
```diff
@@ -0,0 +1,21 @@
+  ignore_tags:
+    - net
+    - bluetooth
+vendor: espressif
```
@kartben, nitpick by @LucasTambor : vendor here should be `m5stack` instead.

kartben replied:
not a nitpick at all - thanks! I will update all other m5stack boards accordingly.

kartben committed:
boards: xtensa: Set M5Stack vendor for M5Stack boards

s/espressif/extensa/ for the couple M5Stack boards in-tree that had the
wrong vendor set.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

marekmatej commented on boards/xtensa/m5stack_atoms3/board.cmake:
```diff
@@ -0,0 +1,9 @@
+if(NOT "${OPENOCD}" MATCHES "^${ESPRESSIF_TOOLCHAIN_PATH}/.*")
+  set(OPENOCD OPENOCD-NOTFOUND)
+endif()
+find_program(OPENOCD openocd PATHS ${ESPRESSIF_TOOLCHAIN_PATH}/openocd-esp32/bin NO_DEFAULT_PATH)
```
Since debugging is not supported, maybe consider reducing this file to single line:
`include(${ZEPHYR_BASE}/boards/common/esp32.board.cmake)`

marekmatej commented on boards/xtensa/m5stack_atoms3/m5stack_atoms3-pinctrl.dtsi:
```diff
@@ -0,0 +1,54 @@
+			pinmux = <SPIM2_MOSI_GPIO21>;
+			output-low;
+		};
+
```
remove blank line

marekmatej reviewed (commented):
some nitpicks, but looks good
